polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -76,26 +76,37 @@ static bool isValidMemRefType(Type type) {
   return true;
 }
 
-// Returns true if the block of \p startOp contains an operation (after \p
-// startOp) that has a side effects on \p val.
+// Returns true if the block of \p startOp contains an operation (after startOp)
+// that has a side effects on \p val, and false otherwise.
 static bool existsSideEffectAfter(Value val, Operation *startOp,
                                   AliasAnalysis &aliasAnalysis) {
   assert(startOp && "Expecting a valid pointer");
-  Block *block = startOp->getBlock();
 
-  for (Operation &op : *block) {
-    if (op.isBeforeInBlock(startOp) || isMemoryEffectFree(&op))
+  for (Operation &op : *startOp->getBlock()) {
+    auto MEI = dyn_cast<MemoryEffectOpInterface>(op);
+    if (op.isBeforeInBlock(startOp) || isMemoryEffectFree(&op) || !MEI)
       continue;
 
-    // Check whether the operation uses an operand that alias with 'val'.
-    if (auto MEI = dyn_cast<MemoryEffectOpInterface>(op)) {
-      if (!llvm::all_of(op.getOperands(), [&](Value operand) {
-            AliasResult aliasRes = aliasAnalysis.alias(operand, val);
-            return aliasRes.isNo();
-          }))
+    // Conservatively assume an operation with a nested region has side effects
+    // on 'val'.
+    if (op.hasTrait<OpTrait::HasRecursiveMemoryEffects>())
+      return true;
+
+    SmallVector<MemoryEffects::EffectInstance, 1> effects;
+    MEI.getEffects(effects);
+    for (MemoryEffects::EffectInstance &effect : effects) {
+      Value effectVal = effect.getValue();
+      if (effectVal == val)
+        return true;
+
```

Comment:
can be removed.

Reply:
Yes because the code at lines 104-106 will return true if `val == effectVal` (they MustAlias). However this is a shortcut that allows to skip the more expensive call to  aliasAnalysis.alias that is the reason I put the early exit.